tls/s2n_protocol_preferences.c

```diff
@@ -17,39 +17,107 @@
+}
 
+S2N_RESULT s2n_protocol_preferences_set(struct s2n_blob *application_protocols, const char *const *protocols, int protocol_count)
+{
+    ENSURE_MUT(application_protocols);
+
+    /* NULL value indicates no preference so free the previous blob */
     if (protocols == NULL || protocol_count == 0) {
-        /* NULL value indicates no preference, so nothing to do */
-        return 0;
+        GUARD_AS_RESULT(s2n_free(application_protocols));
+        return S2N_RESULT_OK;
     }
 
-    GUARD(s2n_stuffer_growable_alloc(&protocol_stuffer, 256));
-    for (int i = 0; i < protocol_count; i++) {
+    DEFER_CLEANUP(struct s2n_blob new_protocols = { 0 }, s2n_free);
+
+    /* Allocate enough space to avoid a reallocation for every entry
+     *
+     * We assume that each protocol is most likely 8 bytes or less.
+     * If it ends up being larger, we will expand the blob automatically
+     * in the append method.
+     */
+    GUARD_AS_RESULT(s2n_realloc(&new_protocols, protocol_count * 8));
+
+    /* set the size back to 0 so we start at the beginning.
+     * s2n_realloc will just update the size field here
+     */
+    GUARD_AS_RESULT(s2n_realloc(&new_protocols, 0));
+
+    for (size_t i = 0; i < protocol_count; i++) {
+        const uint8_t * protocol = (const uint8_t *)protocols[i];
         size_t length = strlen(protocols[i]);
-        uint8_t protocol[255];
 
-        S2N_ERROR_IF(length > 255 || (s2n_stuffer_data_available(&protocol_stuffer) + length + 1) > 65535, S2N_ERR_APPLICATION_PROTOCOL_TOO_LONG);
-        memcpy_check(protocol, protocols[i], length);
-        GUARD(s2n_stuffer_write_uint8(&protocol_stuffer, length));
-        GUARD(s2n_stuffer_write_bytes(&protocol_stuffer, protocol, length));
+        /**
+         *= https://tools.ietf.org/rfc/rfc7301#section-3.1
+         *# Empty strings
+         *# MUST NOT be included and byte strings MUST NOT be truncated.
+         */
+        ENSURE(length < 256, S2N_ERR_INVALID_APPLICATION_PROTOCOL);
+
+        GUARD_RESULT(s2n_protocol_preferences_append(&new_protocols, protocol, (uint8_t)length));
     }
 
-    GUARD(s2n_stuffer_extract_blob(&protocol_stuffer, application_protocols));
-    GUARD(s2n_stuffer_free(&protocol_stuffer));
-    return 0;
+    /* now we can free the previous list since we've validated all new input */
+    GUARD_AS_RESULT(s2n_free(application_protocols));
+
+    /* update the connection/config application_protocols with the newly allocated blob */
+    *application_protocols = new_protocols;
+
+    /* zero out new_protocols so we no longer refer to what we just allocated */
+    new_protocols = (struct s2n_blob){ 0 };
+
+    /* even though we've used DEFER_CLEANUP above, static analysis tools
+     * like cppcheck don't understand that the new assignment will be consumed
+     * at the end of the function. Here we free it even though DEFER_CLEANUP will
+     * also free it. This is OK since s2n_free will be a no-op on empty blobs.
+     */
```

Comment:
If it requires this much explanation, maybe a cppcheck inline suppression would be the better option. We use them elsewhere, and they're pretty straightforward: http://cppcheck.sourceforge.net/manual.pdf#page=16&zoom=auto,-26,442 It would just be:

```
/* cppcheck-suppress unreadVariable */
```

Reply:
I disagree that it would be better. I think including the suppression comment still requires justification as to why it's there.

Reply:
Yes, but it's a shorter explanation. You only have to explain why cppcheck incorrectly thinks there's a problem. So it would just be "cppcheck doesn't recognize that DEFER_CLEANUP uses this variable".

But this isn't a blocker.